### PR TITLE
Aliyun: Add OSSOutputFile

### PR DESCRIPTION
**Pull request by xingbowu**

### CODE REVIEW

File: aliyun/src/main/java/org/apache/iceberg/aliyun/oss/OSSOutputFile.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import org.apache.iceberg.io.OutputFile;
+import org.apache.iceberg.io.PositionOutputStream;
+
+class OSSOutputFile extends BaseOSSFile implements OutputFile {
+
+  OSSOutputFile(OSS client, OSSURI uri, AliyunProperties aliyunProperties) {
+    super(client, uri, aliyunProperties);
+  }
+
+  public static OSSOutputFile fromLocation(OSS client, String location, AliyunProperties aliyunProperties) {
```
Comment (openinx) [inline aliyun/src/main/java/org/apache/iceberg/aliyun/oss/OSSOutputFile.java:35]: As this `OSSOutputFile` is not marked as `public`,  do we need to add the extra `public` for this static method ?

File: aliyun/src/main/java/org/apache/iceberg/aliyun/oss/BaseOSSFile.java [hunk 2]
```diff
@@ -23,16 +23,19 @@
...
 import org.apache.iceberg.relocated.com.google.common.base.MoreObjects;
 
 abstract class BaseOSSFile {
   private final OSS client;
   private final OSSURI uri;
+  private final AliyunProperties aliyunProperties;
   private SimplifiedObjectMeta metadata;
 
-  BaseOSSFile(OSS client, OSSURI uri) {
+  BaseOSSFile(OSS client, OSSURI uri, AliyunProperties aliyunProperties) {
```
Comment (openinx) [inline aliyun/src/main/java/org/apache/iceberg/aliyun/oss/BaseOSSFile.java:35]: Could we keep the original constructor and introduce an extra `BaseOSSFile` constructor with the newly introduced  `aliyunProperties` argument so that we don't need to change the unrelated classes (such as `TestOSSInputFile`, `OSSInputFile` etc) ?

File: aliyun/src/test/java/org/apache/iceberg/aliyun/oss/TestOSSOutputFile.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+  @Test
+  public void testWriteFile() throws IOException {
+    OSSURI uri = randomURI();
+
+    int dataSize = 1024 * 1024 * 10;
+    byte[] data = randomData(dataSize);
+
+    OutputFile out = OSSOutputFile.fromLocation(ossClient, uri.location(), aliyunProperties);
+    try (OutputStream os = out.createOrOverwrite()) {
+      IOUtils.write(data, os);
```
Comment (openinx) [inline aliyun/src/test/java/org/apache/iceberg/aliyun/oss/TestOSSOutputFile.java:54]: Same comment as this one: https://github.com/apache/iceberg/pull/3404#discussion_r739861968

File: aliyun/src/test/java/org/apache/iceberg/aliyun/oss/TestOSSOutputFile.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    OSSURI uri = randomURI();
+
+    int dataSize = 8;
+    byte[] data = randomData(dataSize);
+    writeOSSData(uri, data);
+
+    OutputFile out = OSSOutputFile.fromLocation(ossClient, uri.location(), aliyunProperties);
+    AssertHelpers.assertThrows("Should complain about location already exists", AlreadyExistsException.class,
+        "Location already exists",
+        () -> out.create());
```
Comment (openinx) [inline aliyun/src/test/java/org/apache/iceberg/aliyun/oss/TestOSSOutputFile.java:84]: Nit:  Lambda can be replaced with method reference.

File: aliyun/src/test/java/org/apache/iceberg/aliyun/oss/TestOSSOutputFile.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+    int actualSize = 1024;
+    byte[] actual = randomData(actualSize);
+    OutputFile out = OSSOutputFile.fromLocation(ossClient, uri.location(), aliyunProperties);
+    try (OutputStream os = out.createOrOverwrite()) {
+      IOUtils.write(actual, os);
+    }
+
+    Assert.assertEquals(String.format("Should overwrite object length from %d to %d", dataSize, actualSize),
+        ossClient.getObject(uri.bucket(), uri.key()).getObjectMetadata().getContentLength(), actualSize);
```
Comment (openinx) [inline aliyun/src/test/java/org/apache/iceberg/aliyun/oss/TestOSSOutputFile.java:103]: I think this `ossClient.getObject(uri.bucket(), uri.key()).getObjectMetadata().getContentLength()` can be simplified as `out.toInputFile().getLength()`.

File: aliyun/src/test/java/org/apache/iceberg/aliyun/oss/TestOSSOutputFile.java [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+    int dataSize = 8;
+    byte[] data = randomData(dataSize);
+    writeOSSData(uri, data);
+
+    int actualSize = 1024;
+    byte[] actual = randomData(actualSize);
+    OutputFile out = OSSOutputFile.fromLocation(ossClient, uri.location(), aliyunProperties);
+    try (OutputStream os = out.createOrOverwrite()) {
+      IOUtils.write(actual, os);
```
Comment (openinx) [inline aliyun/src/test/java/org/apache/iceberg/aliyun/oss/TestOSSOutputFile.java:99]: Nit:  Replace the apache common io class with guava's `ByteStream#copy`

File: aliyun/src/test/java/org/apache/iceberg/aliyun/oss/TestOSSOutputFile.java [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+    OSSURI uri = randomURI();
+    OutputFile out = new OSSOutputFile(ossClient, uri, aliyunProperties);
+    Assert.assertEquals("Location should match", uri.location(), out.location());
+  }
+
+  @Test
+  public void testToInputFile() {
+    OutputFile out = new OSSOutputFile(ossClient, randomURI(), aliyunProperties);
+    InputFile in = out.toInputFile();
+    Assert.assertTrue("Should be an instance of OSSInputFile", in instanceof OSSInputFile);
```
Comment (openinx) [inline aliyun/src/test/java/org/apache/iceberg/aliyun/oss/TestOSSOutputFile.java:121]: I think we'd better to assert more informations such as: 
* the location of output file and input file;
* the length of the output file and input file;
* the existence of the input file.

**Review by openinx — COMMENTED** [review]
Thanks @xingbowu for the work, I think we've made a great progress for the aliyun oss integration work.  I left several comments for this !

File: aliyun/src/main/java/org/apache/iceberg/aliyun/oss/BaseOSSFile.java [hunk 8]
```diff
@@ -23,18 +23,25 @@
...
+  private AliyunProperties aliyunProperties;
   private SimplifiedObjectMeta metadata;
 
   BaseOSSFile(OSS client, OSSURI uri) {
     this.client = client;
     this.uri = uri;
   }
 
+  BaseOSSFile(OSS client, OSSURI uri, AliyunProperties aliyunProperties) {
+    this(client, uri);
```
Comment (openinx) [inline aliyun/src/main/java/org/apache/iceberg/aliyun/oss/BaseOSSFile.java:41]: Instead of initializing this constructor on top of the `BaseOSSFile(OSS client, OSSURI uri)`,  I will prefer to initialize the `BaseOSSFile(OSS client, OSSURI uri)` on top of this constructor. In the current approach,  the `aliyunProperties` will be set as a `null`, then we have to check whether it is nullable everywhere in the upper classes such as `OSSOutputStream`. 

I think the aliyunProperties can be initialized as a `new AliyunProperties()` by default.

File: aliyun/src/main/java/org/apache/iceberg/aliyun/oss/OSSOutputFile.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+package org.apache.iceberg.aliyun.oss;
+
+import com.aliyun.oss.OSS;
+import org.apache.iceberg.aliyun.AliyunProperties;
+import org.apache.iceberg.exceptions.AlreadyExistsException;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.io.OutputFile;
+import org.apache.iceberg.io.PositionOutputStream;
+
+public class OSSOutputFile extends BaseOSSFile implements OutputFile {
```
Comment (openinx) [inline aliyun/src/main/java/org/apache/iceberg/aliyun/oss/OSSOutputFile.java:29]: This class don't need to be exposed as `public` because we are only exposed as the abstracted `OutputFile` in iceberg project.

File: aliyun/src/test/java/org/apache/iceberg/aliyun/oss/TestOSSInputFile.java [hunk 10]
```diff
@@ -60,7 +60,8 @@ public void testReadFile() throws Exception {
   public void testOSSInputFile() {
     OSSURI uri = randomURI();
     AssertHelpers.assertThrows("File length should not be negative", ValidationException.class,
-        "Invalid file length", () -> new OSSInputFile(ossClient().get(), uri, -1));
+        "Invalid file length",
+        () -> new OSSInputFile(ossClient().get(), uri, -1));
```
Comment (openinx) [inline aliyun/src/test/java/org/apache/iceberg/aliyun/oss/TestOSSInputFile.java:64]: Nit: Pls revert this unrelated change.

File: aliyun/src/test/java/org/apache/iceberg/aliyun/oss/TestOSSOutputFile.java [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+    OutputFile out = OSSOutputFile.fromLocation(ossClient, uri.location(), aliyunProperties);
+    try (OutputStream os = out.create()) {
+      ByteStreams.copy(is, os);
+    }
+
+    Assert.assertTrue("OSS file should exist", ossClient.doesObjectExist(uri.bucket(), uri.key()));
+    Assert.assertEquals("Object length should match", ossDataLength(uri), dataSize);
+
+    byte[] actual = new byte[dataSize];
+    ByteStreams.readFully(ossDataContent(uri), actual);
```
Comment (openinx) [inline aliyun/src/test/java/org/apache/iceberg/aliyun/oss/TestOSSOutputFile.java:61]: Looks like we are forgetting to close the input stream from http connection, right ?

Comment (openinx) [reply aliyun/src/test/java/org/apache/iceberg/aliyun/oss/TestOSSOutputFile.java:61]: Maybe we can just return the byte array from `ossDataContent` and close it inside the `ossDataContent`.

**Review by openinx — COMMENTED** [review]
Left other comments for the second round review.

**Comment by xingbowu** [comment]
Thanks a lot for reviewing this PR in time @openinx , I have addressed them.

**Review by openinx — APPROVED** [review]
LGTM, Thanks @xingbowu for the contribution !
